addons/account_asset/account_asset.py

```diff
@@ -1,351 +1,325 @@
-        'history_ids': fields.one2many('account.asset.history', 'asset_id', 'History', readonly=True),
-        'depreciation_line_ids': fields.one2many('account.asset.depreciation.line', 'asset_id', 'Depreciation Lines', readonly=True, states={'draft':[('readonly',False)],'open':[('readonly',False)]}),
-        'salvage_value': fields.float('Salvage Value', digits_compute=dp.get_precision('Account'), help="It is the amount you plan to have that you cannot depreciate.", readonly=True, states={'draft':[('readonly',False)]}),
-    }
-    _defaults = {
-        'code': lambda obj, cr, uid, context: obj.pool.get('ir.sequence').next_by_code(cr, uid, 'account.asset.code'),
-        'purchase_date': lambda obj, cr, uid, context: time.strftime('%Y-%m-%d'),
-        'active': True,
-        'state': 'draft',
-        'method': 'linear',
-        'method_number': 5,
-        'method_time': 'number',
-        'method_period': 12,
-        'method_progress_factor': 0.3,
-        'currency_id': lambda self,cr,uid,c: self.pool.get('res.users').browse(cr, uid, uid, c).company_id.currency_id.id,
-        'company_id': lambda self, cr, uid, context: self.pool.get('res.company')._company_default_get(cr, uid, 'account.asset.asset',context=context),
-    }
-
-    def _check_recursion(self, cr, uid, ids, context=None, parent=None):
-        return super(account_asset_asset, self)._check_recursion(cr, uid, ids, context=context, parent=parent)
-
-    def _check_prorata(self, cr, uid, ids, context=None):
-        for asset in self.browse(cr, uid, ids, context=context):
-            if asset.prorata and asset.method_time != 'number':
-                return False
         return True
 
-    _constraints = [
-        (_check_recursion, 'Error ! You cannot create recursive assets.', ['parent_id']),
-        (_check_prorata, 'Prorata temporis can be applied only for time method "number of depreciations".', ['prorata']),
-    ]
-
-    def onchange_category_id(self, cr, uid, ids, category_id, context=None):
-        res = {'value':{}}
-        asset_categ_obj = self.pool.get('account.asset.category')
+    @api.multi
+    def validate(self):
+        self.write({'state': 'open'})
+
+    @api.multi
+    def set_to_close(self):
+        unposted_dep_line = self.env['account.asset.depreciation.line'].search_count(
+            [('asset_id', 'in', self.ids), ('move_check', '=', False)])
+        if unposted_dep_line:
+            raise UserError(_('You cannot close a document which has unposted lines.'))
+        self.message_post(body=_("Document closed."))
+        self.write({'state': 'close'})
+
+    @api.multi
+    def set_to_draft(self):
+        self.write({'state': 'draft'})
+
+    @api.one
+    @api.depends('depreciation_line_ids.move_check')
```

Comment:
probably `@api.depends('depreciation_line_ids')` would be appropriate, because you will probably want to recalculate the residual value if the line amount is changed...